pkg/api/types.go

```diff
@@ -1509,10 +1513,26 @@ type NodeList struct {
 	Items []Node `json:"items"`
 }
 
-// NamespaceSpec describes the attributes on a Namespace
+// NamespaceNetworkPolicy determines who is authorized to access pods in the namespace
+type NamespaceNetworkPolicy string
+
+// These are the valid network policies of a namespace
+const (
+	// Closed namespaces are only accessible by pods within the namespace
+	NamespaceNetworkPolicyClosed NamespaceNetworkPolicy = "Closed"
+	// Open namespaces are accessible from any namespace
+	NamespaceNetworkPolicyOpen NamespaceNetworkPolicy = "Open"
+)
+
+// NamespaceSpec describes the attributes on a Namespace.
 type NamespaceSpec struct {
-	// Finalizers is an opaque list of values that must be empty to permanently remove object from storage
-	Finalizers []FinalizerName
+	// Finalizers is an opaque list of values that must be empty to permanently remove object from storage.
+	// More info: http://releases.k8s.io/HEAD/docs/design/namespaces.md#finalizers
+	Finalizers []FinalizerName `json:"finalizers,omitempty"`
+
+	// NetworkPolicy indicates who is authorized to access pods in the namespace.
+	// Must be either Open or Closed. Defaults to Open.
+	NetworkPolicy NamespaceNetworkPolicy `json:"experimentalNetworkPolicy,omitempty"`
```

Comment:
Also think this should be a separate type that is added to a namespace
versus a field on the namespace.  Annotation on namespace is fine for
experimentation, but not final API.  I think in a future where policy is
more core to Kube (like we have in OpenShift) it would be a different role
to change network policy versus edit other fields on a namespace.  For
example, if a namespace has a description that users want to edit, I should
not need to also give them ability to edit the entire network policy.

On Friday, November 13, 2015, Derek Carr <decarr@redhat.com> wrote:

> Also agree experimental features are best exposed via annotations similar
> to what was done with bandwidth limiting via a pod annotation.  It should
> not be a field in the API yet.
>
> Also recording a note to ensure this feature can be disabled completely
> when I have more time to review such that even if the annotation was
> specified it was ignored.
>
> On Friday, November 13, 2015, Brian Grant <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> In pkg/api/types.go
>> <https://github.com/kubernetes/kubernetes/pull/13937#discussion_r44849563>
>> :
>>
>> >  type NamespaceSpec struct {
>> > -	// Finalizers is an opaque list of values that must be empty to permanently remove object from storage
>> > -	Finalizers []FinalizerName
>> > +	// Finalizers is an opaque list of values that must be empty to permanently remove object from storage.
>> > +	// More info: http://releases.k8s.io/HEAD/docs/design/namespaces.md#finalizers
>> > +	Finalizers []FinalizerName `json:"finalizers,omitempty"`
>> > +
>> > +	// NetworkPolicy indicates who is authorized to access pods in the namespace.
>> > +	// Must be either Open or Closed. Defaults to Open.
>> > +	NetworkPolicy NamespaceNetworkPolicy `json:"experimentalNetworkPolicy,omitempty"`
>>
>> So far, we have not added policies directly to namespace. If this is
>> experimental, why not use an annotation?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/13937/files#r44849563>.
>>
>